### ISSUE TITLE
API to receive mouse back/forward button events (mouse buttons 4 and 5)

### ISSUE THREAD

**Issue by dweymouth** [fyne-io/fyne#3812]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same feature, before opening a new one.
- [X] This issue only relates to a single feature. I will open new issues for any other features.

### Is your feature request related to a problem?

There is currently no way to react to mouse button events other than the first three buttons and the scroll wheel.

### Describe the solution you'd like to see.

Introduce a new API for responding to additional mouse button events. New constants should be added to the [`desktop.MouseButton`](https://github.com/fyne-io/fyne/blob/master/driver/desktop/mouse.go) enum representing the back/forward mouse buttons (button numbers 4 and 5). It's unclear what the names should be. `MouseButtonBack/Forward` is inflexible, but continuing the existing naming scheme is also awkward (`MouseButtonQuaternary`, `MouseButtonQuinary`)?

Also unclear is where the events should be surfaced. A natural option would be through the existing `desktop.Mouseable` interface and the `MouseEvent` type, but a potential complication is that the mouse back/forward buttons are usually best responded to at the window level, and with this API any existing Mouseable CanvasObject would "swallow" the event, even if it can't meaningfully respond to it, preventing it from reaching the top-level window canvas. Perhaps exposing a new interface `desktop.ExtendedMouseable` or something like that so that existing Mouseable objects wouldn't receive them, and then adding new `SetOnMouseDown`, `SetOnMouseUp` functions to the desktop Canvas that receive any mouse events (from the normal or extended buttons) that were not sent to a CanvasObject.

**Comment by andydotxyz** [comment]
Is this a standard? My understanding was that buttons >3 were mapped to OS functions by the mouse driver.

This is a nice addition idea, but it's implementation/naming needs to match a naming / function scheme that works across all OS.

**Comment by dweymouth** [comment]
I *think* though I could be wrong that by default those buttons are unmapped and just issue mouseButton events that it's up to the application to handle (which is mapped to forward/back in web browsers and other apps with navigation).  But the user can override them at the driver level to have it issue a key press event or other event type to remap the behavior.

I can try to dig a bit more into this later this week

**Comment by andydotxyz** [comment]
More info would be good - I could not find any consensus online. If we have to expose button numbers so be it, but a consistent intent api would be better.

**Comment by dweymouth** [comment]
In Win32, mouse buttons 4 and 5 are delivered to the application to respond to: https://learn.microsoft.com/en-us/windows/win32/learnwin32/mouse-clicks

Cocoa dispatches them with NSEventTypeOtherMouseDown and then has a buttonNumber property in the payload: https://stackoverflow.com/questions/69030392/how-to-differentiate-side-and-wheel-buttons-with-cocoa

Qt has APIs for responding to buttons >3 as well: https://stackoverflow.com/questions/30822134/qml-forwards-back-mouse-buttons-handling

So I think it's clear that these events, by default, are dispatched to applications, though many mouse drivers allow them to be remapped. So I think Fyne should have an API to receive these. I like the ExtendedMouseable idea to keep existing Mouseable widgets that don't respond to these from "swallowing" the event and preventing it to reach the window canvas.

**Comment by andydotxyz** [comment]
Yeah. It seems that the intent is not something we can rely on and may just have to fall back to numbering.
Where they should land is a tougher one though, and probably impacts how we should implement them too. The "delivered to canvas" is more like a shortcut than a mouse action... and maybe that is what it is? But then again some applications (CAD etc) like to use mouse taps in the context of the widget to control items in the app (like zoom, pan etc?).
I wonder if there is any sort of best practice we can pull on here, or perhaps we need the "ExtendedMouseable" to be available on a desktop.Window or desktop.Canvas as well as widgets?

**Comment by dweymouth** [comment]
Another idea is a new interface that widgets can implement to specify which mouse buttons they can receive, and default for widgets that don't implement it is they receive buttons 1, 2, and 3 only (as is currently the case). Then in the future, in 3.0 we could merge the two interfaces together so any Mouseable widget now has to say which buttons it's interested in

**Comment by andydotxyz** [comment]
I think that approach only works for 3.0 as in the 2.x if you implement the interface you handle the events - no event handling is optional once you opt in at the code level...

**Comment by andydotxyz** [comment]
This doesn't feel very clean. If you are simply tappable but the thing you are on top of has a popup menu it would be undesirable for a right-click on the child to show the pop-up associated with the widget you are not currently interacting with.

It feels more like these "meta buttons" are special in some way - are they back/forward in the app or are they "apply a fourth action" on a widget. Looking at it that way I wonder if they are more like modifiers or keys than mouse buttons.
Whichever way we go I don't think that the API design should break how mouse interactions currently work.

**Comment by dweymouth** [comment]
They are *mostly* back/forward within the app but I'm not sure you can make the assumption that they would always be tied to that functionality (games, etc).

But I have no rush for getting this out in a release I guess since it would be easy for me to add temporary support for Supersonic only in a fork.

**Comment by pjanx** [comment]
Many mice are rather reprogrammable and can send keyboard events. It is safe to assume that buttons 4 and 5 are for history use, but you can theoretically also have more buttons than that (reported as mouse buttons in terms of USB HID).

**Comment by andydotxyz** [comment]
>  It is safe to assume that buttons 4 and 5 are for history use

If it is programmable then we certainly /cannot/ assume that they will always be history

**Comment by pjanx** [comment]
I think I haven't made myself clear: the two left physical buttons can (often) be reprogrammed to send different events than buttons 4 and 5, such as keys A or B.

**Comment by dweymouth** [comment]
It's still not safe to assume buttons 4 and 5 are for history use, because applications that don't have history may want to map those button events to some other action (eg games, or something with graphical shape manipulation)

**Comment by andydotxyz** [comment]
This is marked as a blocker, can we figure out how to solve the API? Is going with the numbered buttons the only save approach?

**Comment by pjanx** [comment]
Note to also merge the numbering across systems.

Linux/X11 uses buttons like this: left, middle, right, scroll up, scroll down, scroll left, scroll right, history back, history forward, *.
(I don't actually own a mouse with horizontal scrolling, but I assume it will be in the 6, 7 space. Logitech makes these.)

**Comment by andydotxyz** [comment]
A good point, thanks - we certainly cannot just call them button 4 and 5 then, as on Linux they will be 6 and 7 because the scroll wheel counts where it doesn't on Windows and macOS.

Perhaps instead `MouseExtraButton1` (and 2, 3 etc?)
Then the OS can set these as appropriate constants so the API reads consistently?

**Comment by dweymouth** [comment]
That naming works I suppose, and I also think `MouseButtonForward`/`MouseButtonBack` works as well since that is the most common use of these buttons, and it of course doesn't necessitate that a custom widget or mouse handler must treat them this way.

I think more important is the requirement that these additional mouse buttons shouldn't get dispatched to "legacy" `Mouseable` and `Tappable` implementers (like widget.Button, etc) since this would cause such widgets to swallow the events without responding to them and block the most common use of these buttons which is reacting to them at a window- or container-level for back/forward navigation. That, to me, implies a new interface like `ExtendedMouseable` where widgets specify which "extra" mouse button events they are interested in receiving.

**Comment by andydotxyz** [comment]
> That, to me, implies a new interface like ExtendedMouseable 

Hmm, Maybe - but the Button is already a field on `Mouseable` so developers would be `switch`ing it already

> where widgets specify which "extra" mouse button events they are interested in receiving.

Not sure about this one - that is a huge departure from any event types we have. It would be moving to a `Shortcut` model instead, and I'm not clear that we have the need for such a change.

Either way, clearly not in the sort of quick change I was hoping for in the last days of 2.8 feature work - pushing.
